conans/client/settings_preprocessor.py

```diff
@@ -1,30 +1,73 @@
+import warnings
+
+from conans.client.build.cppstd_flags import cppstd_default
 from conans.client.build.cppstd_flags import cppstd_flag
 from conans.errors import ConanException
 from conans.util.log import logger
 
 
 def preprocess(settings):
-    fill_runtime(settings)
-    check_cppstd(settings)
+    _fill_runtime(settings)
+    _fill_compiler_cppstd(settings)
+    _check_cppstd(settings)
+
+
+def _fill_compiler_cppstd(settings):
+    compiler = settings.get_safe("compiler")
+    compiler_version = settings.get_safe("compiler.version")
+    cppstd = settings.get_safe("cppstd")
+    compiler_cppstd = settings.get_safe("compiler.cppstd")
+
+    # Assign the explicit default value to compiler.cppstd (only if not given cppstd)
+    if not cppstd and not compiler_cppstd and compiler and compiler_version:
+        default_cppstd = cppstd_default(compiler, compiler_version)
+        if default_cppstd:
+            try:
+                settings.compiler.cppstd = default_cppstd
+            except Exception:
+                # Settings structure does not have cppstd
+                pass
 
 
-def check_cppstd(settings):
+def _check_cppstd(settings):
     compiler = settings.get_safe("compiler")
     compiler_version = settings.get_safe("compiler.version")
     cppstd = settings.get_safe("cppstd")
-    if not cppstd or compiler not in ("gcc", "clang", "apple-clang", "Visual Studio"):
+    compiler_cppstd = settings.get_safe("compiler.cppstd")
+
+    if not cppstd and not compiler_cppstd:
         return
-    cpp_values = settings.cppstd.values_range
-    available = [v for v in cpp_values if cppstd_flag(compiler, compiler_version, v)]
-    if str(cppstd) not in available:
-        raise ConanException("The specified 'cppstd=%s' is not available "
-                             "for '%s %s'. Possible values are %s'" % (cppstd,
-                                                                       compiler,
-                                                                       compiler_version,
-                                                                       available))
+
+    # Checks: one or the other, but not both
+    if cppstd and compiler_cppstd:
+        raise ConanException("Do not use settings 'compiler.cppstd' together with 'cppstd'."
+                             " Use only the former one.")
+
+    if cppstd:
+        warnings.warn("Setting 'cppstd' is deprecated in favor of 'compiler.cppstd'")
```

Comment:
I'm curious, why not `self.output.warn`?

Reply:
This raises a `UserWarning` if you have configured _warnings as errors_ (we do in the test suite), if not it just prints the message.